components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -7,83 +7,108 @@
+// Contract Application Binary Interface (ABI).
 //
 // Method arguments are packed in chunks of 32 bytes, with types smaller than
-// 32 bytes zero padded if necessary. Fixed-size types like uint256, address
-// are represented within the available 32 bytes, however, dynamic types like
-// bytes and arrays follow head-tail encoding. In this scheme, the data is
-// packaged at the tail-end of the transaction's calldata. The arguments are
-// references into the calldata where the content is.
+// 32 bytes zero padded if necessary. Fixed-size types like uint256, address,
+// bytes32, etc., are represented within the available 32 bytes, however,
+// dynamic types like bytes and arrays follow head-tail encoding. In this
+// scheme, the data is packaged at the tail-end of the transaction's calldata.
+// The arguments are references into the calldata where the content is.
 //
-// Parsing of complex cases such as composite types, nested arrays, etc., are
-// not implemented. The decoder only extracts the calldata tail reference in
-// such cases.
+// The decoder will parse the calldata and return a base::Value object, which
+// can be serialized to JSON or used in other contexts. If the decoder
+// encounters an unknown type or an error, it will return an empty optional.
+//
+// Unsupported types:
+// - int<M>
+// - fixed<M>x<N>
+// - ufixed<M>x<N>
+// - function
 //
 // Function selector (first 4 bytes) should NOT be part of the calldata being
 // parsed.
 //
 // References:
-//   https://docs.soliditylang.org/en/latest/abi-spec.html
+//   - https://docs.soliditylang.org/en/latest/abi-spec.html
+//   -
+//   https://github.com/web3/web3.js/tree/4.x/packages/web3-eth-abi/src/coders/base
 
 namespace brave_wallet {
 
 namespace {
-// GetArgFromData extracts a 32-byte wide hex string from the calldata at the
-// specified offset. The parsed value is NOT prefixed by "0x".
-std::optional<std::string> GetArgFromData(const std::vector<uint8_t>& input,
-                                          size_t offset) {
-  if (offset > input.size() || input.size() - offset < 32) {
-    return std::nullopt;
+
+using ByteArray = std::vector<uint8_t>;
+
+// DecoderResult is a tuple that contains the decoded value, the remaining
+// calldata, and the number of bytes consumed.
+template <typename T>
+using DecoderResult = std::tuple<T, ByteArray, size_t>;
+
+constexpr size_t kWordSize = 32;
+constexpr size_t kAddressSize = 20;
+
+// GetSubByteArray returns a subarray of the input stas rting from the specified
+// offset. If the offset is out of bounds, an empty ByteArray is returned.
+//
+// It mimics the behavior of JavaScript's Uint8Array.prototype.subarray.
+ByteArray GetSubByteArray(const ByteArray& input, size_t offset) {
+  if (offset >= input.size()) {
+    return ByteArray();
   }
 
-  return HexEncodeLower(input.data() + offset, 32);
+  return ByteArray(input.begin() + offset, input.end());
 }
 
```

Comment:
This approach might be considered as too heavy in terms of number of allocations. 
It is ok to allocate output values, but not ok to allocate for input data. Hierarchical parsing should be done by narrowing spans